From acab99640f8a349e82a04a29ea6afdf13f941912 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Thu, 30 Aug 2018 15:46:56 +0800
Subject: [PATCH 4538/5242] MLK-19413-27 gpu: imx: dpu: extdst: Add
 extdst_pixengcfg_syncmode_master() helper

commit  f2cabde17f717b9c9dd85aec9e09058d2db3e45a from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds extdst_pixengcfg_syncmode_master() helper support
so that the callers may control if a extdst is master or slave
when it works in sync mode.  The bit16 of extdst's PIXENGCFG_STATIC
register controls this and it's a part of sync mode fixup logic.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-extdst.c |   19 +++++++++++++++++++
 include/video/dpu.h              |    1 +
 2 files changed, 20 insertions(+)

diff --git a/drivers/gpu/imx/dpu/dpu-extdst.c b/drivers/gpu/imx/dpu/dpu-extdst.c
index 6c6555b..23a1b8a 100644
--- a/drivers/gpu/imx/dpu/dpu-extdst.c
+++ b/drivers/gpu/imx/dpu/dpu-extdst.c
@@ -178,6 +178,25 @@ void extdst_pixengcfg_div(struct dpu_extdst *ed, u16 div)
 }
 EXPORT_SYMBOL_GPL(extdst_pixengcfg_div);
 
+void extdst_pixengcfg_syncmode_master(struct dpu_extdst *ed, bool enable)
+{
+	struct dpu_soc *dpu = ed->dpu;
+	u32 val;
+
+	if (!dpu->devtype->has_syncmode_fixup)
+		return;
+
+	mutex_lock(&ed->mutex);
+	val = dpu_pec_ed_read(ed, PIXENGCFG_STATIC);
+	if (enable)
+		val |= BIT(16);
+	else
+		val &= ~BIT(16);
+	dpu_pec_ed_write(ed, val, PIXENGCFG_STATIC);
+	mutex_unlock(&ed->mutex);
+}
+EXPORT_SYMBOL_GPL(extdst_pixengcfg_syncmode_master);
+
 int extdst_pixengcfg_src_sel(struct dpu_extdst *ed, extdst_src_sel_t src)
 {
 	struct dpu_soc *dpu = ed->dpu;
diff --git a/include/video/dpu.h b/include/video/dpu.h
index f679d88..e7184d2 100644
--- a/include/video/dpu.h
+++ b/include/video/dpu.h
@@ -529,6 +529,7 @@ void constframe_constantcolor(struct dpu_constframe *cf, unsigned int r,
 void extdst_pixengcfg_sync_mode(struct dpu_extdst *ed, ed_sync_mode_t mode);
 void extdst_pixengcfg_reset(struct dpu_extdst *ed, bool reset);
 void extdst_pixengcfg_div(struct dpu_extdst *ed, u16 div);
+void extdst_pixengcfg_syncmode_master(struct dpu_extdst *ed, bool enable);
 int extdst_pixengcfg_src_sel(struct dpu_extdst *ed, extdst_src_sel_t src);
 void extdst_pixengcfg_sel_shdldreq(struct dpu_extdst *ed);
 void extdst_pixengcfg_shdldreq(struct dpu_extdst *ed, u32 req_mask);
-- 
1.7.9.5

